src/indieweb_utils/feeds/discovery.py

```diff
@@ -67,3 +68,41 @@ def discover_web_page_feeds(url: str, user_mime_types: Optional[List[str]] = Non
         feeds.append(FeedUrl(url=url, mime_type="text/html", title=page_title.text))
 
     return feeds
+
+
+def discover_h_feed(url: str) -> dict:
+    """
+    Find the main h-feed that represents a web page as per the h-feed Discovery algorithm.
+
+    refs: https://microformats.org/wiki/h-feed#Discovery
+
+    :param url: The URL of the page whose associated feeds you want to retrieve.
+    :type url: str
+    :return: The h-feed data.
+    :rtype: dict
+    """
+
+    parsed_main_page_mf2 = mf2py.parse(url=url)
+
+    all_page_feeds = discover_web_page_feeds(url)
+
+    if all_page_feeds.get("text/mf2+html"):
+        # url has already been canonicalized by discover_web_page_feeds
+        # thus, no canonicalization is needed in this function
```

Comment:
I think we could use this function independently of discover_web_page_feeds, so if we can run canonicalization on an already canonicalized url and get the same thing back, it's better to just try and re-canonicalize it so we're not making any assumptions about how this function will be used.

Reply:
Can you elaborate on this suggestion? The discover_web_page_feeds function retrieves all of the feeds on a page, including the mf2 feed that someone is likely not to have discovered yet.

With regard to re-canonicalization, good call! I shall ad dspuport for an additional call to canonicalize_url in my next PR.